### ISSUE TITLE
Optional auto-generation of operator== and hashCode on classes with const constructors

### ISSUE THREAD

**Issue by Hixie** [dart-lang/sdk#26703]
It is a common Dart idiom to have a class with all-final fields, a const constructor, a toString, an operator==, and a hashCode function.

The latter two are boilerplate:
```
  @override
  bool operator ==(dynamic other) {
    if (other.runtimeType != runtimeType)
      return false;
    final T typedOther = other;
    return typedOther.field1 == field1
        && typedOther.field2 == field2
        && typedOther.fieldN == fieldN;
  }

  @override
  int get hashCode => hashValues(field1, field2, fieldN);
```

It would be nice if there was some syntax or shorthand we could provide that would automatically provide these.

cc @abarth

**Comment by a14n** [comment]
Something like https://github.com/a14n/zengen#value :)

**Comment by anders-sandholm** [comment]
Thanks, @a14n, for the link to https://pub.dartlang.org/packages/zengen 
@Hixie, @abarth, I'd be curious to know what you think about this.

**Comment by abarth** [comment]
Quite close, but the generated code isn't exactly what we'd want.  We'd want

```dart
@DoTheThing()
class A {
  final a;
  final int b;
}
```

to generate

```dart
class A {
  final a;
  final int b;
  const A({this.a, this.b});
  @override bool operator ==(o) => identical(this, o) ||
      o.runtimeType == runtimeType && o.a == a && o.b == b;
  @override int get hashCode => hashObjects(a, b);
}
```

Also, we don't want to add a source-to-source translation step to our toolchain.  We'd like something that the compiler and analyzer understand.

**Comment by yjbanov** [comment]
@abarth I think the proposed desugaring can only work cleanly under the following conditions:

* `Object` is `A`'s direct super-class
* `A` is sealed/final (cannot be extended or implemented)
* The types of all fields of `A` satisfy these same three conditions

Otherwise, if `A` is allowed to participate in any non-trivial class hierarchy, then the proposed implementation of `operator==` will violate Liskov substitution principle. For example, `CartesianPoint { x; y }` will not be `operator==` equal to `CylindricalPoint { unitVector; radius; }` even if they both implement the same `Point` interface.

**Comment by abarth** [comment]
Is there a better `operator==` function we should use?  Here's a real one:

```dart
  @override
  bool operator ==(dynamic other) {
    assert(debugAssertIsValid());
    if (identical(this, other))
      return true;
    if (other is! BoxConstraints)
      return false;
    final BoxConstraints typedOther = other;
    assert(typedOther.debugAssertIsValid());
    return minWidth == typedOther.minWidth &&
           maxWidth == typedOther.maxWidth &&
           minHeight == typedOther.minHeight &&
           maxHeight == typedOther.maxHeight;
  }
```

https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/rendering/box.dart#L404

We'd be willing to use the `debugAssertIsValid` if we didn't have to actually write out the overload.

**Comment by Hixie** [comment]
@yjbanov A `CartesianPoint` isn't equal to a `CylindricalPoint`, they have different types.

**Comment by Hixie** [comment]
Incidentally, in Flutter's codebase we sometimes check for type equality using `is` and sometimes using `runtimeType`. There are some specific cases where `is` is the right answer because we want the Liskov substitution principle to apply (e.g. `_DebugSize` vs `Size`) but usually we actually want `runtimeType` equality. I've been slowly moving us towards the latter, but there's lots of cases where we still use the former because we didn't know better when we last did our big pass at making our `operator ==`s consistent.

The reason you want to use `operator ==` is that otherwise `foo == bar` might give a different result than `bar == foo` (in particular in the case where `foo` is an instance of a subclass of the class `bar` is an instance of, and `foo`'s class introduces new fields, `foo == bar` might be false even when `bar == foo` is true).

**Comment by lrhn** [comment]
@Hixie 

I see that as a clear sign that `operator==` is badly designed in general. (It's not the only sign).
It's just one of those functions that shouldn't be virtual to begin with, because overriding it in a subclass is  pretty much impossible without breaking semantics. 

If we had non-virtual methods, I would make `operator==` one and then a `Set<Point>` should only distinguish points by coordinates, even if some are actually `ColorPoint` instances. From the *set*'s perspective, they are the same `Point`.
A `Set<ColorPoint>` would distinguish by colors, but it also only sees `ColorPoint` instances.
Then you could override `operator==` to only accept objects of the correct type, instead of having to accept `Object` and always do the is-check first and reject everything anyway.

(Another approach is to not have your instances extend each other, but have all of them extend a common abstract superclass, with equality only being defined at the leaves of the class tree, then you can use `is` checks, and still allow mocks to successfully emulate the instance classes. Basically: Never put an `operator==` implementation on a class that can be extended).

**Comment by Hixie** [comment]
We use `operator ==` on complicated type hierarchies a lot. It works fine so long as you're testing the runtimeType for equality.

**Comment by yjbanov** [comment]
@abarth 

> Is there a better operator== function we should use?

I suspect there's no one implementation of `operator==` that's better than others in all situations, and that's the basis of my concern: if one particular implementation is baked into the language it may come back and bite us later. Such syntactic sugar provides an incentive to use it even in cases where it's not the right thing to do.

> we don't want to add a source-to-source translation step to our toolchain. We'd like something that the compiler and analyzer understand.

What is the difference between a source-to-source translator and a compiler? It seems they both need to do the same amount of work and generate the same thing.

@Hixie 

> A `CartesianPoint` isn't equal to a `CylindricalPoint`, they have different types.

Types don't matter in my case. Both classes express the same thing. The choice between one type or the other is purely an implementation detail (perhaps they have different performance characteristics). If `runtimeType`s are compared, this implementation detail will leak to the user.

**Comment by abarth** [comment]
> What is the difference between a source-to-source translator and a compiler? It seems they both need to do the same amount of work and generate the same thing.

Maybe a more correct thing to say is that we want something that's fully integrated into the Dart toolchain and has zero impact on the development cycle.  The existing source-to-source transformations don't have these properties but macro expansion in C, for example, does have these properties.

**Comment by Hixie** [comment]
If you're ok with adding a `CartesianPoint` to a `Set` and getting back a `CylindricalPoint`, then implementing your own `operator ==`/`hashCode` pair that does that should definitely be possible.

This bug is only about how to provide an `operator ==`/`hashCode` pair that does a straight-forward comparison of all the final fields of an immutable class (one with all-final fields, a const constructor, a toString, an operator==, and a hashCode function). We do this all the time in Flutter's codebase, and it would be nice to not have to worry about writing this boilerplate code all the time. We've also received requests for this from our customers.

**Comment by Hixie** [comment]
Oh certainly. As I noted above, even in Flutter's codebase we don't always want the same version. For example equivalent Size and _DebugSize objects want to be considered ==, whereas equivalent Offset and Size objects do not (these all directly or indirectly inherit from OffsetBase). All we're looking for here is a shorthand for a common case we see a lot. It may be that the common case here isn't common enough to justify dedicated syntax and instead there should be macro-like syntax. Or it could be that the common case here really is very common and it can be blessed, while still allowing less common cases to be implemented on a case-by-case basis.

**Comment by yjbanov** [comment]
I vaguely remember seeing a macro-like proposal that would rely on annotations to generate code. IIRC, the use-case was to generate JSON serializers/deserializers from fields and some sort of AOP use-case. Can't find it anymore. I think it was @sigmundch's doc.

**Comment by a14n** [comment]
https://github.com/dart-lang/dart_enhancement_proposals/blob/f433a56ba95f4750f0c795a1b89323be3ff3f1cc/Meetings/2015-05-06%20DEP%20Committee%20Meeting.md#interceptors-and-macros ?

**Comment by yjbanov** [comment]
Bingo! Go, @a14n!

**Comment by jmesserly** [comment]
for what it's worth -- C# 7 is going to have this.
https://github.com/dotnet/roslyn/blob/features/records/docs/features/records.md

**Comment by yjbanov** [comment]
@jmesserly Makes sense as (if I'm reading correctly) they do this only for sealed classes and structs, both of which satisfy the conditions [above](#issuecomment-226684512)

**Comment by jmesserly** [comment]
Yup, exactly. From the looks of it they designed record types to make immutable plain-old-data classes very easy & they also thought about how it integrates with pattern matching (https://github.com/dotnet/roslyn/blob/features/records/docs/features/patterns.md)

**Comment by davidmorgan** [comment]
FYI

Here is a source_gen based approach to this problem:

https://github.com/google/built_value.dart 

source_gen satisfies the requirement to be visible to the analyzer.

built_value is modeled after Guava's @AutoValue. Re: the equality operator, we follow this from @AutoValue: https://github.com/google/auto/blob/master/value/userguide/howto.md#inherit -- and point to Effective Java, which explains that there is no good way to mix equality and inheritance.

**Comment by abarth** [comment]
> source_gen satisfies the requirement to be visible to the analyzer.

Can you explain this a bit more?

**Comment by davidmorgan** [comment]
> > source_gen satisfies the requirement to be visible to the analyzer.

> Can you explain this a bit more?

source_gen uses the build package:

https://github.com/dart-lang/source_gen
https://github.com/dart-lang/build

The idea behind "build" is that when developing you run it as a watcher than continuously updates the generated source.

So from the point of view of your IDE and the analyzer there's no difference between generated and non-generated source.

It's a nice experience, you can navigate easily into the generated code and see what's going on.

**Comment by abarth** [comment]
Thanks.  Unfortunately, that doesn't meet my requirement to have zero impact on the development cycle.  It sounds like a good approach to have a small impact, but it's still nonzero.

**Comment by davidmorgan** [comment]
Yes, it's not quite as neat as a language feature.

I've given considerable thought to the possibilities around a language feature. Right now I think the source_gen approach is better.

If we add something as a language feature, it becomes very fixed/rigid, we can't iterate on it. So my intention is to build with source_gen in the hope of settling on something widely successful/popular. Then there will be a case for upgrading it to a language feature.

What I want is more than is described on this request, though. For an immutable class I like to also have a builder class, so you have a mutable version of the type that you can accumulate data in. This is good for readability and efficiency. This is what built_value, linked above, provides. I don't see a nice way to do this with a language feature.

**Comment by abarth** [comment]
> Right now I think the source_gen approach is better.

The source_gen approach is not better for us because it does not meet our requirements.

**Comment by davidmorgan** [comment]
Thanks. I understand that source_gen does not meet your requirements. What I meant by "better" is in the short term: the source_gen approach is a realistic way to create interest/support for something before it becomes a language feature.

Put another way, if we first have a popular source_gen solution to the problem, there is a much stronger case for adding it as a language feature. I would also like to see something end up in the language at some point, but I'd be surprised if it happens soon, so that's why I'm working on the source_gen path.

**Comment by abarth** [comment]
You're welcome to head down the source_gen approach, but it's somewhat off-topic for this bug.

**Comment by mockturtl** [comment]
@yjbanov @a14n 
> I vaguely remember seeing a macro-like proposal

dart-lang/dart_enhancement_proposals#44

**Comment by tatumizer** [comment]
@lrhn 
Your argument rather shows that equality is not an internal property of the object at all (be it a virtual operator or otherwise). It's a property of "viewpoint" on the collection of objects - so it cannot be hardcoded in object itself. It's only when we define Set<ColorPoint>, then it's clear what equality means from the point of view of this specific collection. For Set<Point> equality is different.

If equality is not a property of object, neither is hashCode

Apparently, correct solution is to generate == and hashCode outside of the object - in collection, depending on the type of collection. So, object can have several "external" == and hashCode implementations.

**Comment by tatumizer** [comment]
BTW, there's no other way to guarantee == is commutative.

**Comment by lrhn** [comment]
Yes, equality (or, to be exact, "equivalence") is not a property of the object. but of the perspective.
In mathematics, equality is really identity. If you work on semantic values, a point is a member of ℝ². You can't have "two points" that are both at (2, 2) - then it is the *same* point.

In Object Oriented programming, objects often correspond to "real" concepts and entities. That's why you can have a Cartesian point (x=2, y=2) and a polar point (r= √8, φ= π/4) which both represent the same underlying entity - the element (2, 2) of ℝ².
If you look at those objects merely as points, they are "mathematically equal" - they represent the same entity. If both implement a `Point` interface, there is no reason to not accept them as equal `Point`s - they are "equal modulo representation".

On the other hand, if the representation is important to you, you can have a function that expects either a `CartesianPoint` or a `PolarPoint`, and then the two are not equivalent to you. Your perspective has changed from the represented entity to its representation.

And even further, looking at the actual objects in the system, even two `CartesianPoint`s with the same coordinates are not `identical`. That's the finest granularity of distinction allowed by the language.
Objects have identity, that's a fundamental tenet of object orientation. That's both a blessing and a curse - it allows you to *refer* to the object (references only make sense if the thing you refer to is a specific thing). It allows mutability - you can't *change* something if it has no identity. It also prevents some optimizations because you can't be certain that you are the only one who can access an object - anyone with a reference can access the same object.

So, why do we have equals on some objects anyway.

For one thing, even if there are multiple perspectives on an object, in some cases one perspective is so much more common than any other that it makes sense to give it precedence. Numerical equality of numbers. Contents of strings. It's simply too annoying not to be able to compare two numbers by just writing `n == 4` - the alternative is `numericalEquality(n, 4)`, and it just doesn't roll of your tongue the same way.
You can still declare an equality on integers that is `equals modulo 2**32`, it's just not the "natural" equality, and you have to represent it separately.

When you have an *interface*, it makes sense to declare a notion of equality on that interface, based only on the visible properties. That makes it possible for multiple implementations to agree on the equality. If you implement the interface by extending a base-class, you can even share the equality implementation. The `num` interface declares an equality that is shared by `int` and `double`.

You don't have to give a class a custom equality. Many classes work well with just the default identity - objects are equal only if they are the same objects (that's really the minimal equivalence relation: it is the minimal reflexive relation and that happens to also be trivially symmetric and transitive).
Dart does not have equality on, say, `List`. We could have, but we have tried to not put equality on mutable objects. That's dangerous, and not always useful. 
It does mean that we lose out on the cases where easily comparing lists of integers would be practical, but that's a trade-off, not just an oversight.

So, back to equality on classes that can be extended.
It's hard. 
If your subclass is significantly different from the superclass (e.g., `Point` vs `ColorPoint`) then you might not want to make them equal. However, a `ColorPoint` is-a `Point`, and from a `Point` perspective it might be equal to another point which isn't colored. Then you get either non-symmetry when a `Point` considers itself equal to a `ColorPoint` with the same coordinates, but not the other way around, or non-transitivity if the `ColorPoint` do consider itself equal to a plain `Point` with the same coordinates, but not equal to another `ColorPoint` with the same coordinates and a different color.
It's just broken.
Basically, if you need to overwrite the equality in a subclass, then it's probably not really an equality on the superclass's interface  any more. It's two different equalities, and they are not necessarily compatible, and by putting them on the objects, you loose the ability to pick the perspective you want.

Using `runtimeType` to disallow equality with any subclass is, well, possible and consistent and prevents accidentally being equal to a `ColorPoint`, but it also breaks substitutability at its core and prevents you from being equal to a `PolarPoint`. I can no longer create a specialization of the class or a different implementation of the interface that can be used in all places where the superclass can. (Well, you can, because we allow you to override `runtimeType` too and lie about your type - yey?)

In summary: Don't put equals on mutable objects or objects likely to be extended. Only put it at the leaves of your class hierarchy, or pick one equality and use it for the entire subtree. And even then it will sometimes suck.

**Comment by tatumizer** [comment]
@lrhn: I'm not sure we have any disagreement on the issue, Just a couple of nitpicks.

The only type that has clearly defined equality is int (IMO). Double doesn't. Cartesian point (x=2, y=2) and a polar point (r= √8, φ= π/4) might be the same in some "philosophical" sense, but they are not the same for computer - e.g. the latter, after conversion to Cartesian coordinates, prints (2.0, 2.0000000000000004).

That's why, in the context of any computational algorithm, you never see comparison of doubles for equality - instead, some epsilon-interval is used. I think equality of doubles is useful only when these doubles are actually integer values, and programmer is aware of that.

Equality of strings is debatable, too. E.g., Strings can be equal in swift, but not equal in dart - see "Comparing strings" in [1]. Indeed, like you said, it would be very inconvenient to NOT have equality for ints, and it's very inconvenient to have swiftian equality for strings. "Convenience" is a key word. It should be a key word in the context of any discussion we have on any topic.

The bottom line of my entire tirade: equality is a very fuzzy notion, even in "simple" cases. 
 
[1] https://developer.apple.com/library/ios/documentation/Swift/Conceptual/Swift_Programming_Language/StringsAndCharacters.html

**Comment by tatumizer** [comment]
What if we try to explore a different idea: Comparator\<T>(listOfFields).
E.g. Comparator\<Point>([\#x, \#y]). or Comparator\<ColorPoint>(\#x,\#y, \#color) etc

One good property of such comparator is: it can be autogenerated. There's enough information in this "constructor call" to generate everything: equals(), hashCode, even compare() for inequalities.

That is, you are not supposed to write any code - you write Comparator\<Point>([\#x, \#y]) - and compiler does the rest automatically. If list is omitted, it can by default use all final fields or something.

Comparator has many other potentially useful properties, but I want to keep the post short.

Certainly, to use it, collections have to provide optional parameter to pass your comparator.
I think the concept is equivalent to "user-defined perspective" we discussed above.

How about that?

**Comment by davidmorgan** [comment]
Re: Comparator<Point>, I think the answer is the same: you can do this right now with codegen, so it doesn't make sense to pursue a language feature.

**Comment by tatumizer** [comment]
> you can do *this* right now with codegen,
It depends on what you mean by "this". You cannot generate Comparator\<Point> - instead, it will be something like PointComparator. Further, because in general, there can be several comparators for the same class (depending on perspective) generated class will be called PointComparatorByXAndY, the beauty of which is questionable.

In general, dart doesn't have a coherent generation concept right now. Generated code often mixes user-defined code with generated code, (What if you need 2 generators transforming the same class? It becomes even more confusing). In a word, you won't get "this" - what you get is something different.

**Comment by davidmorgan** [comment]
I expect a source_gen based solution would look reasonable.

You _can_ generate `Comparator<Point>`; using source_gen you declare the interface you want then delegate to generated code. So, yes, the generated code will be `PointComparator` -- actually, `_$PointComparator` -- but you'll never see it, you'll see `Comparator<Point>`:

```
part 'point.g.dart';

class Point {
  @Compare(#x)
  static final Comparator<Point> byX = _$byX;
  @Compare(#x, #y)
  static final Comparator<Point> byXandY = _$byXandY;

  int x;
  int y;
}
```

source_gen supports multiple generators on one file, no trouble.

**Comment by tatumizer** [comment]
Hm... That's passable. Not as cool, but probably OK. So, problem solved? Or what?

**Comment by davidmorgan** [comment]
I hope so :) ... I think it's good enough for many problems, what's needed is for people to try it out and explore what works in practice. If we can suggest small language changes that make the codegen story better, that seems a good direction.

**Comment by tatumizer** [comment]
As a minimum, dart needs to support Comparator interface, with equals(), compareTo() and hashCode.
And it has to accept it an optional parameter in Map, Set and similar. (BTW, at the time of writing. term "Comparator" is already defined in dart core, and means something different). Also, because we need hashCode there, maybe a better name for the class needs to be figured out. 

It's not really a language change though - rather, backwards-compatible API change.

**Comment by tatumizer** [comment]
I'd like to mention one nice property of Comparator as defined above.

Suppose you have a bunch of objects of the same type, and want to create an index - e.g. by fields x and y. That is, you need a map (or multimap) that takes tuple [x, y] and maps it to object(s). 

Standard way of doing it is to use Map, but then you need to form keys as separate objects. Whatever representation you choose for tuple [x, y] - concatenated Strings or special "Key" object with fields x, y, or something else - this key is a full-fledged object. Sometimes it takes almost as much memory as original object (due to overhead). So you effectively double the amount of memory. And if you need another index, by different set of fields, - you triple it, To make things worse, memory affects performance (due to cache misses). Useless keys pollute Space and Time.

When you use Comparator, you don't need separate Key objects - equals and hashCode are defined by comparator directly based on original objects.

**Comment by Hixie** [comment]
Assuming the comparator is reified at compile time, I don't see why it wouldn't take the same amount of memory as today.

**Comment by tatumizer** [comment]
Because with Comparator, you don't need keys. E.g. if you have objects indexed by combination of String fields x and y, you currently do something like map[obj.x+'#'+obj.y]=obj. Key  is a string obj.x+'#'+obj.y - takes memory.
OK, you may define special object for key e.g. MyKey(x,y) and create MyKey(obj.x, obj.y) - but it's still an object, with all overhead incurred by object.

With Comparator, you don't need any extra objects.

**Comment by Hixie** [comment]
I don't understand. How would it work with Comparator then?

**Comment by tatumizer** [comment]
Assuming map provides extra parameter to pass our comparator, I create map
Map map=new Map(pointComparator) // see definition of pointComparator  above
then I call
for (obj in myPoints) map[obj]=obj; 
Note that I pass the *obj* as a key, so no new objects get created. Because comparator is based on fields x and y only, it doesn't matter how many other fields Point has - the key will effectively be a pair (x, y).

If you need to search for Point with coordinates 2, 2, create Point(2, 2) and use it as key

**Comment by Hixie** [comment]
Oh, I see, you're basically passing in a method to the `Map` to do the comparison instead of using `==` and `hashCode`.

That doesn't help our general case, unfortunately. We have lots of classes in the Flutter framework that compare objects and rely on the `operator ==` result, and thus lots of objects that have `operator ==` implementations. This issue is just about requesting a way to do those without all the boilerplate we have today. If you want `Map` extended as you describe, I think that should be a separate issue.

**Comment by tatumizer** [comment]
In your case, instead of calling == directly, you can call myComparator.equals(point1, point2)
And you can have as many comparators as you wish. You can even include runtimeType into comparator (not sure under what scenario this is helpful, but nothing will stop you from defining
Comparator\<Point>([ \#x, \#y, \#runtimeType])

**Comment by Hixie** [comment]
Sure. We can do that today, no need for any new syntax or anything. It's significantly uglier than just having an `==` that does the right thing though.

**Comment by tatumizer** [comment]
You can express == in terms of generated Comparator directly in your class. Then no one in the outside world will know you have a Comparator inside. There's still boilerplate, but you can't expect to get rid of all boilerplate completely: somehow, you have to declare that you wish to include "runtimeType" in equation :)

**Comment by tatumizer** [comment]
Forget to mention: writing == and hashCode manually might be not trivial. Pitfalls are briefly listed here: https://commons.apache.org/proper/commons-lang/javadocs/api-2.6/org/apache/commons/lang/builder/EqualsBuilder.html

So maybe using generator (internally) wouldn't be such a bad idea even for your case, despite boilerplate? Not sure.

**Comment by Hixie** [comment]
That page just says it's hard (and then gives some Java-specific issues). Our experience is not that it's hard, only that it's tedious.

Ideally the solution wouldn't involve writing code for both hashCode and ==. Exactly how it is implemented isn't something I'm particularly concerned about, so long as it is fast and the syntax is brief, convenient, and readable.

**Comment by tatumizer** [comment]
Went in great lengths in search of better name for "Comparator". After hours of googling, found something similar: "HashingStrategy"! Try to guess the language where it belongs! I don't know, either the author had a good sense of humor, or that's what passes for Strategy these days, but it's not a good fit for dart.

Came up with a couple of ideas of my own: 1) Equarator 2) (even better) Equator. 
How about that?  Please don't hesitate to send your suggestions!

**Comment by lrhn** [comment]
In `dart:collection` we (read: I) used `Equality`. Not perfect, it's not obvious that it includes the hash code, but it really is an implementation of an equality relation.

**Comment by tatumizer** [comment]
> In dart:collection I used Equality

In package:collection. You see how confusing it is? 

BTW, Comparer (term used by C#) is not a word either - at least, it's unknown to Webster.
And "hash", for Webster, means something else already.

Anyway, in package:collection I can't find "Index" class that allows creation of index by combination of fields (as discussed above). I don't want to hijack this thread - are you OK if I open another issue for this, in package:collection github repository?

**Comment by lrhn** [comment]
On Wed, Aug 31, 2016 at 4:41 PM, Tatumizer <notifications@github.com> wrote:

> In dart:collection I used Equality
>
> In package:collection. You see how confusing it is?
>

Doh! :)


> BTW, Comparer (term used by C#) is not a word either - at least, it's
> unknown to Webster.
>
It's a derived word - it's something that compares. There is also the
comparees that the comparer compares.


> And "hash", for Webster, means something else already.
>
> Anyway, in package:collection I can't find "Index" class that allows
> creation of index by combination of fields (as discussed above). I don't
> want to hijack this thread - are you OK if I open another issue for this,
> in package:collection github repository?
>
Go ahead. If it's something that can be done without reflection, I don't
expect to have a problem with it.

/L
-- 
Lasse R.H. Nielsen - lrn@google.com
'Faith without judgement merely degrades the spirit divine'
Google Denmark ApS - Frederiksborggade 20B, 1 sal - 1360 København K
- Denmark - CVR nr. 28 86 69 84
